core/src/main/java/com/linecorp/armeria/internal/DefaultAttributeMap.java

```diff
@@ -155,19 +178,34 @@
                 }
                 curr = curr.next;
             }
-            return false;
+            return hasAttrInRoot(key);
         }
     }
 
-    /**
-     * Returns all {@link Attribute}s this map contains.
-     */
+    private <T> boolean hasAttrInRoot(AttributeKey<T> key) {
+        if (root != null) {
+            return root.hasAttr(key);
+        }
+        return false;
+    }
+
+    @Override
     public Iterator<Attribute<?>> attrs() {
         final AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
         if (attributes == null) {
-            return Collections.emptyIterator();
+            return root != null ? root.attrs() : Collections.emptyIterator();
```

Comment:
I just realized we need to copy on demand or copy all attributes here because otherwise the attribute changes will affect the root.